File: 22_专题/RMQ问题/SparseTable2D.go
// 二维RMQ/二维ST表
// https://codeforces.com/blog/entry/45485
// Preprocess : O(n*m*logn*logm)
//  we create a table[ 1+logn ][ n ][ 1+logm ][ m ]
//  Each box of the table[ 1+logn ][ n ] is a sparse table of size [ 1+logm ][ m ]
//  Let us see what table[ jr ][ ir ][ jc ][ ic ] actually contains:
//  It contains the minimum element from column ic to ic-1+2^jc of all rows from ir to ir-1+2^jr
//  In other words, it contain the minimum element in the submatrix [ (ir,ic), (ir-1+2^jr , ic-1+2^jc) ]
//  where submatrix [ (x1,y1),(x2,y2) ] denotes the submatrix with x1,y1 as its top left-most and x2,y2 as its bottom right-most point.
//  Now you can easily conclude that, table[ 0 ][ ir ][ jc ][ ic ] is nothing but the 1D RMQ table if we take our array as row ir

package main

import (
	"fmt"
	"math/bits"
)

func main() {
	matrix := [][]int{
		{1, 2, 3, 4, 5},
		{6, 7, 8, 9, 10},
		{11, 12, 13, 14, 15},
	}

	query := NewSparseTable2D(matrix, max)
	fmt.Println(query(1, 1, 2, 2))
}

// 3148. 矩阵中的最大得分
// https://leetcode.cn/problems/maximum-difference-score-in-a-grid/description/
// 给你一个由 正整数 组成、大小为 m x n 的矩阵 grid。你可以从矩阵中的任一单元格移动到另一个位于正下方或正右侧的任意单元格（不必相邻）。从值为 c1 的单元格移动到值为 c2 的单元格的得分为 c2 - c1 。
// 你可以从 任一 单元格开始，并且必须至少移动一次。
// 返回你能得到的 最大 总得分。

const INF int32 = 1e9 + 10

func maxScore(grid [][]int) int {
	grid32 := make([][]int32, len(grid))
	for i := range grid {
		grid32[i] = make([]int32, len(grid[i]))
		for j := range grid[i] {
			grid32[i][j] = int32(grid[i][j])
		}
	}

	ROW, COL := int32(len(grid32)), int32(len(grid32[0]))
	res := -INF
	st := NewSparseTable2D(grid32, min32)
	for i := int32(0); i < ROW; i++ {
		for j := int32(0); j < COL; j++ {
			// up
			if i > 0 {
				upMin := st(0, j, i-1, j)
				res = max32(res, grid32[i][j]-upMin)
			}
			// left
			if j > 0 {
				leftMin := st(i, 0, i, j-1)
				res = max32(res, grid32[i][j]-leftMin)
			}

			// left and up
			if i > 0 && j > 0 {
				upLeftMin := st(0, 0, i-1, j-1)
				res = max32(res, grid32[i][j]-upLeftMin)
			}
		}
	}

	return int(res)
}

// query: 查询 [row1,col1,row2,col2] 闭区间的贡献值
//
//	0 <= row1 <= row2 < len(matrix)
//	0 <= col1 <= col2 < len(matrix[0])
func NewSparseTable2D[S any](matrix [][]S, op func(S, S) S) (query func(row1, col1, row2, col2 int32) S) {
	n, m := int32(len(matrix)), int32(len(matrix[0]))
	rowSize := int32(bits.Len32(uint32(n))) // 1+logn
	colSize := int32(bits.Len32(uint32(m))) // 1+logm
	dp := make([][][][]S, rowSize)          // (rowSize * row * colSize * col)
	for i := int32(0); i < rowSize; i++ {
		dp[i] = make([][][]S, n)
		for j := int32(0); j < n; j++ {
			dp[i][j] = make([][]S, colSize)
			for k := int32(0); k < colSize; k++ {
				dp[i][j][k] = make([]S, m)
			}
		}
	}

	for ir := int32(0); ir < n; ir++ {
		for ic := int32(0); ic < m; ic++ {
			dp[0][ir][0][ic] = matrix[ir][ic]
		}
		for jc := int32(1); jc < colSize; jc++ {
			for ic := int32(0); ic+(1<<jc) <= m; ic++ {
				dp[0][ir][jc][ic] = op(dp[0][ir][jc-1][ic], dp[0][ir][jc-1][ic+(1<<(jc-1))])
			}
		}
	}

	for jr := int32(1); jr < rowSize; jr++ {
		for ir := int32(0); ir+(1<<jr) <= n; ir++ {
			for jc := int32(0); jc < colSize; jc++ {
				for ic := int32(0); ic+(1<<jc) <= m; ic++ {
					dp[jr][ir][jc][ic] = op(dp[jr-1][ir][jc][ic], dp[jr-1][ir+(1<<(jr-1))][jc][ic])
				}
			}
		}
	}

	query = func(row1, col1, row2, col2 int32) S {
		rowk := bits.Len32(uint32(row2-row1+1)) - 1
		colK := bits.Len32(uint32(col2-col1+1)) - 1
		res1 := op(dp[rowk][row1][colK][col1], dp[rowk][row1][colK][col2-(1<<colK)+1])
		res2 := op(dp[rowk][row2-(1<<rowk)+1][colK][col1], dp[rowk][row2-(1<<rowk)+1][colK][col2-(1<<colK)+1])
		return op(res1, res2)
	}

	return
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min32(a, b int32) int32 {
	if a < b {
		return a
	}
	return b
}

func max32(a, b int32) int32 {
	if a > b {
		return a
	}
	return b
}

File: 22_专题/RMQ问题/SparseTable2D.ts
// 二维RMQ/二维ST表
// https://codeforces.com/blog/entry/45485
// Preprocess : O(n*m*logn*logm)

class SparseTable2D<S> {
  private readonly _e: () => S
  private readonly _op: (a: S, b: S) => S
  private readonly _dp: S[]
  private readonly _row: number
  private readonly _col: number
  private readonly _hash1: number
  private readonly _hash2: number
  private readonly _hash3: number

  constructor(matrix: ArrayLike<ArrayLike<S>>, e: () => S, op: (a: S, b: S) => S) {
    const row = matrix.length
    const col = matrix[0].length
    const rowSize = 32 - Math.clz32(row)
    const colSize = 32 - Math.clz32(col)
    const dp: S[] = Array(rowSize * row * colSize * col)
    for (let i = 0; i < dp.length; i++) dp[i] = e()

    const hash1 = row * colSize * col
    const hash2 = colSize * col
    const hash3 = col

    for (let ir = 0; ir < row; ir++) {
      for (let ic = 0; ic < col; ic++) {
        dp[ir * hash2 + ic] = matrix[ir][ic]
      }
      for (let jc = 1; jc < colSize; jc++) {
        for (let ic = 0; ic + (1 << jc) <= col; ic++) {
          dp[ir * hash2 + jc * hash3 + ic] = op(
            dp[ir * hash2 + (jc - 1) * hash3 + ic],
            dp[ir * hash2 + (jc - 1) * hash3 + ic + (1 << (jc - 1))]
          )
        }
      }
    }

    for (let jr = 1; jr < rowSize; jr++) {
      for (let ir = 0; ir + (1 << jr) <= row; ir++) {
        for (let jc = 0; jc < colSize; jc++) {
          for (let ic = 0; ic + (1 << jc) <= col; ic++) {
            dp[jr * hash1 + ir * hash2 + jc * hash3 + ic] = op(
              dp[(jr - 1) * hash1 + ir * hash2 + jc * hash3 + ic],
              dp[(jr - 1) * hash1 + (ir + (1 << (jr - 1))) * hash2 + jc * hash3 + ic]
            )
          }
        }
      }
    }

    this._e = e
    this._op = op
    this._dp = dp
    this._row = row
    this._col = col
    this._hash1 = hash1
    this._hash2 = hash2
    this._hash3 = hash3
  }

  /**
   * 查询闭区间`[row1,col1,row2,col2]`的贡献值.
   * 0 <= row1 <= row2 < len(matrix).
   * 0 <= col1 <= col2 < len(matrix[0]).
   */
  query(row1: number, col1: number, row2: number, col2: number): S {
    if (row1 < 0) row1 = 0
    if (col1 < 0) col1 = 0
    if (row2 >= this._row) row2 = this._row - 1
    if (col2 >= this._col) col2 = this._col - 1
    if (row1 > row2 || col1 > col2) return this._e()
    const rowK = 32 - Math.clz32(row2 - row1 + 1) - 1
    const colK = 32 - Math.clz32(col2 - col1 + 1) - 1
    const { _hash1, _hash2, _hash3, _dp, _op } = this
    const res1 = _op(
      _dp[rowK * _hash1 + row1 * _hash2 + colK * _hash3 + col1],
      _dp[rowK * _hash1 + row1 * _hash2 + colK * _hash3 + col2 - (1 << colK) + 1]
    )
    const res2 = _op(
      _dp[rowK * _hash1 + (row2 - (1 << rowK) + 1) * _hash2 + colK * _hash3 + col1],
      _dp[
        rowK * _hash1 + (row2 - (1 << rowK) + 1) * _hash2 + colK * _hash3 + col2 - (1 << colK) + 1
      ]
    )
    return _op(res1, res2)
  }
}

class SparseTable2DInt32 {
  private readonly _e: () => number
  private readonly _op: (a: number, b: number) => number
  private readonly _dp: Int32Array
  private readonly _row: number
  private readonly _col: number
  private readonly _hash1: number
  private readonly _hash2: number
  private readonly _hash3: number

  constructor(
    matrix: ArrayLike<ArrayLike<number>>,
    e: () => number,
    op: (a: number, b: number) => number
  ) {
    const row = matrix.length
    const col = matrix[0].length
    const rowSize = 32 - Math.clz32(row)
    const colSize = 32 - Math.clz32(col)
    const dp = new Int32Array(rowSize * row * colSize * col).fill(e())

    const hash1 = row * colSize * col
    const hash2 = colSize * col
    const hash3 = col

    for (let ir = 0; ir < row; ir++) {
      for (let ic = 0; ic < col; ic++) {
        dp[ir * hash2 + ic] = matrix[ir][ic]
      }
      for (let jc = 1; jc < colSize; jc++) {
        for (let ic = 0; ic + (1 << jc) <= col; ic++) {
          dp[ir * hash2 + jc * hash3 + ic] = op(
            dp[ir * hash2 + (jc - 1) * hash3 + ic],
            dp[ir * hash2 + (jc - 1) * hash3 + ic + (1 << (jc - 1))]
          )
        }
      }
    }

    for (let jr = 1; jr < rowSize; jr++) {
      for (let ir = 0; ir + (1 << jr) <= row; ir++) {
        for (let jc = 0; jc < colSize; jc++) {
          for (let ic = 0; ic + (1 << jc) <= col; ic++) {
            dp[jr * hash1 + ir * hash2 + jc * hash3 + ic] = op(
              dp[(jr - 1) * hash1 + ir * hash2 + jc * hash3 + ic],
              dp[(jr - 1) * hash1 + (ir + (1 << (jr - 1))) * hash2 + jc * hash3 + ic]
            )
          }
        }
      }
    }

    this._e = e
    this._op = op
    this._dp = dp
    this._row = row
    this._col = col
    this._hash1 = hash1
    this._hash2 = hash2
    this._hash3 = hash3
  }

  /**
   * 查询闭区间`[row1,col1,row2,col2]`的贡献值.
   * 0 <= row1 <= row2 < len(matrix).
   * 0 <= col1 <= col2 < len(matrix[0]).
   */
  query(row1: number, col1: number, row2: number, col2: number): number {
    if (row1 < 0) row1 = 0
    if (col1 < 0) col1 = 0
    if (row2 >= this._row) row2 = this._row - 1
    if (col2 >= this._col) col2 = this._col - 1
    if (row1 > row2 || col1 > col2) return this._e()
    const rowK = 32 - Math.clz32(row2 - row1 + 1) - 1
    const colK = 32 - Math.clz32(col2 - col1 + 1) - 1
    const { _hash1, _hash2, _hash3, _dp, _op } = this
    const res1 = _op(
      _dp[rowK * _hash1 + row1 * _hash2 + colK * _hash3 + col1],
      _dp[rowK * _hash1 + row1 * _hash2 + colK * _hash3 + col2 - (1 << colK) + 1]
    )
    const res2 = _op(
      _dp[rowK * _hash1 + (row2 - (1 << rowK) + 1) * _hash2 + colK * _hash3 + col1],
      _dp[
        rowK * _hash1 + (row2 - (1 << rowK) + 1) * _hash2 + colK * _hash3 + col2 - (1 << colK) + 1
      ]
    )
    return _op(res1, res2)
  }
}

export { SparseTable2D, SparseTable2DInt32 }

if (require.main === module) {
  const st2d = new SparseTable2DInt32(
    [
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9]
    ],
    () => 0,
    Math.max
  )

  console.log(st2d.query(1, 2, 1, 2))
}
